clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -308,4 +312,42 @@ public void testMetadataFetchOnStaleMetadata() throws Exception {
         PowerMock.verify(metadata);
     }
 
+    @Test
+    public void testTopicRefreshInMetadata() throws Exception {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(ProducerConfig.MAX_BLOCK_MS_CONFIG, "600000");
+        KafkaProducer<String, String> producer = new KafkaProducer<>(props, new StringSerializer(), new StringSerializer());
+        long refreshBackoffMs = 500L;
+        long metadataExpireMs = 60000L;
+        final Metadata metadata = new Metadata(refreshBackoffMs, metadataExpireMs, true, new ClusterResourceListeners());
+        final Time time = new MockTime();
+        MemberModifier.field(KafkaProducer.class, "metadata").set(producer, metadata);
+        MemberModifier.field(KafkaProducer.class, "time").set(producer, time);
+        String topic = "topic";
+
+        Thread t = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                for (int i = 0; i < 12; i++) {
+                    metadata.update(Cluster.empty(), Collections.<String>emptySet(), time.milliseconds());
+                    time.sleep(60 * 1000L);
+                    try {
+                        Thread.sleep(100);
```

Comment:
Are you suggesting that we should call `metadata.update` only once after `metadata.updateRequested` returns true? I am not sure this is better. I think the test needs call metadata.update() multiple times so that the elapsed time is long enough for the topic to be removed from metadata without this patch, but not if we have this bug fix. This also mimics the scenarios that will cause problem. 

If I didn't get your suggestion correctly, can you be more specific what you suggest?